### DISCUSSION TITLE
NextJS Website Breaks On Production But Works Fine On Development

### DISCUSSION THREAD

**Discussion by b-codex** [vercel/next.js#41125] [Help] [answered]
I have been developing an application using NextJS with Typescript and has encountering this problem. I am not new, but I have never encountered this problem while I have done many projects like this and hosted them.

So, in simple words, I created a horizontal menu (main menu) that has contents inside of it. But when i switch around on the main menu, the app basically breaks (this occurs when it's hosted and works without errors when I run it locally). Although navigating around the side menus works fine.

Hosted Website Link: https://inflow-temp.vercel.app/

Demo Video Link: https://www.awesomescreenshot.com/video/11478402?key=c4d65a557445434a7823230cb3cbbfb0

Clone to check it locally. Repo Link: https://gitlab.com/ettadevs/inflow-temp

![image](https://user-images.githubusercontent.com/61845339/193618716-6120ce46-1aec-49a3-90a4-811c28f67805.png)

I am not fetching data from the internet nor mapping anything. I changed it dynamic routing hoping that would fix it, but still encountering the problem on production(hosted).

I have also tried hosting it on netlify just in case if it was vercel.com that was causing the problem, and yet I have had no luck in fixing this problem.

I have scoured the internet looking for answers, but found none.

I don't understand what is causing the problem or how to fix it. Thanks in advance for all your help.

**Comment by icyJoseph** [comment, marked as answer]
So the issue was `swcMinify`. That sounds like, it should be reported as a bug though.

**Reply by b-codex** [reply]
yea. I got an answer on stackoverflow after spending days trying to figure out the problem.
I removed swcMinify: true from next.config.js and it is working fine after that.

**Reply by icyJoseph** [reply]
Well, for what is worth I've isolated the issue to:

```js
function registerSize(key: React.Key, width: number | null) {
    setItemWidths(origin => {
      const clone = new Map(origin);

      if (width === null) {
        clone.delete(key);
      } else {
        clone.set(key, width);
      }
      return clone;
    });
  }
```

Which is to be found within https://github.com/react-component/overflow/blob/dd3fe551fa4b3da1b75735998c21bba98c609d3c/src/Overflow.tsx#L212, but I am yet not 100% sure whose fault it is, you see that `new Map(origin)` call looks very suspicious.

The chain of dependencies is also quite long, `antd`'s Menu uses `rc-menu`, which uses `rc-overflow`, which uses `rc-resize-observer`, which uses `resize-observer-polyfill`, which in turn defines `Map` and `WeakMap`, shims, which means they don't apply if the environment supports them, but I though SWC was failing at minifying that, it did not.

Turns out the problem is within the `rc-overflow` library.  I'll continue looking around.

Also I think the problem is that `new Map(null)` is somehow converted to `new Map(0)`, but I cannot prove it 100% just yet... Gotta learn to use `rc-overflow` 😅

**Reply by icyJoseph** [reply]
~Or perhaps I could skip the rc-overflow library altogether and test the minification of `new Map(null)` directly.~ Nope that wasn't it.

**Reply by b-codex** [reply]
good luck with that. let us know if you manage to fix the problem with "swcMinify: true" flag set on next.config.js.

**Reply by icyJoseph** [reply]
The issue is really awkward, you see, for some reason the state created by Overflow resolves to `[empty, null, Map(0), 0, 0, 375]`, when doing `swcMinify` - but it should be `[null, Map(0), 0, 0, 375]` when running debug it doesn't happen either! 

And then the state reading they have is offset by one, so when it should read `Map(0)` it reads `0`. I cannot boil down what could be possibly happening to trigger the hole at the beginning of the array. 

There's quite a bit of moving parts,  but [`useBatchFrameState`](https://github.com/react-component/overflow/blob/dd3fe551fa4b3da1b75735998c21bba98c609d3c/src/hooks/useBatchFrameState.tsx) defined in `rc-overflow` is where things are going wild though.

**Reply by icyJoseph** [reply]
Alright, I've got my case:

```js
function foo() {
  let walker = 0;

  let arr = [];

  function bar(defaultValue) {
    const myIndex = walker;
    walker += 1;

    console.log({ arr });

    if (arr.length < myIndex + 1) {
      arr[myIndex] = defaultValue;
    }
  }

  return bar;
}

const bar = foo();

bar(null);
bar(null);
bar(null);
bar(null);
bar(null);
```

When compiled with `swc` using this `.swcrc` file:

```json
{
  "jsc": {
    "parser": {
      "syntax": "ecmascript"
    },
    "target": "es5",
    "minify": {
      "compress": true,
      "mangle": true
    }
  }
}
```

Outputs code that causes the walker to `+1` before the `if`-statement that assigns the first default value. So for the first item we are setting the index 1, which creates a whole on the array, and from there it is madness.

```js
var l = (function () {
  var l = 0,
    n = [];
  return function (u) {
    (l += 1),
      console.log({
        arr: n,
      }),
      n.length < l + 1 && (n[l] = u);
  };
})();
l(null), l(null), l(null), l(null), l(null);
```

The final log outputs: `{ arr: [ <1 empty item>, null, null, null, null ] }`, but since the very beginning there's a hole at the beginning of the array.

`package.json` used:

```json
{
  "name": "swc-test",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "devDependencies": {
    "@swc/cli": "^0.1.57",
    "@swc/core": "^1.3.4"
  }
}
```

If we set ` "compress": false,` on the `.swcrc` file, then the bug goes away.

What does this mean for Next.js? Well that perhaps a new version with a fixed version of SWC is needed to be able to use the minifier. I'll try to report the bug to SWC as best as I can so that they can fix it.

In the case of OP:

```js
     var t;
                return a += 1,
                o.current.length < a + 1 && (o.current[a] = e),
                [o.current[a], function(e) {
                    o.current[a] = "function" == typeof e ? e(o.current[a]) : e,
                    M.Z.cancel(l),
                    l = (0,
                    M.Z)(function() {
                        i({}, !0)
                    })
                }
                ]
            }
```

This is the offending code, which is minified from `rc-overflow`, and it contains the `a+=1` statement before the length check right under, which means that `a` is already at value 1 when that check happens, and that creates a whole in the array, which offsets the state reader and causes `new Map(0)` at runtime.

**Reply by icyJoseph** [reply]
https://github.com/swc-project/swc/issues/6039 - A fix was merged! I guess over the next couple of days, or with a new minor release the Next.js will update the package.

**Reply by b-codex** [reply]
thanks for all the help.
